### ISSUE TITLE
SPDistributedCacheService: throws error "Host is not running in cluster" in SharePoint 2019

### ISSUE THREAD

**Issue by Yvand** [PowerShell/SharePointDsc#995]
SPDistributedCacheService: throws error "Host is not running in cluster" in SharePoint 2019

#### Details of the scenario you tried and the problem that is occurring

Resource SPDistributedCacheService throws an error when it configures the service for the 1st time. If I run the DSC configuration again on the server there is no error.
This issue always reproduces on SharePoint 2019, while the exact same DSC configuration works perfectly fine on SharePoint 2016 and 2013.
It can be reproduced in my [ARM template](https://github.com/Yvand/AzureRM-Templates/tree/dev/Templates/SharePoint-ADFS).

#### Verbose logs showing the problem

What puzzles me is that according to [SharePoint log I attached here](https://github.com/PowerShell/SharePointDsc/files/2689800/DistributedCacheError_SP2019.log), it fails on operation StopCacheHost:

```
12/18/2018 09:45:35.09	wsmprovhost.exe (0x09D8)	0x1230	SharePoint Foundation	DistributedCache	aauxg	Unexpected	Unexpected error while executing StopCacheHost with parameters provider: 'SPDistributedCacheClusterProvider' , connectionString: 'Data Source=SQLAlias;Initial Catalog=SPDSC_Config;Integrated Security=True;Persist Security Info=False;Enlist=False;Pooling=True;Min Pool Size=0;Max Pool Size=100;PoolBlockingPeriod=Auto;Asynchronous Processing=False;Connection Reset=True;MultipleActiveResultSets=False;Replication=False;Connect Timeout=15;Encrypt=False;TrustServerCertificate=False;Load Balance Timeout=0;Packet Size=8000;Type System Version=Latest;Application Name=".Net SqlClient Data Provider";User Instance=False;Context Connection=False;Transaction Binding="Implicit Unbind";ApplicationIntent=ReadWrite;MultiSubnetFailover=False;TransparentNetworkIPResolution=True;ConnectRetryCount=1;ConnectRetryInterval=10;Column Encryption Setting=Disabled' , hostName: 'SP' , cachePort: '22233'. Exception: 'System.Management.Automation.CmdletInvocationException: Error occurred while performing the operation on host SP:22233 : ErrorCode<ERRCAdmin016>:SubStatus<ES0001>:Host is not running in cluster. ---> Microsoft.ApplicationServer.Caching.DataCacheException: ErrorCode<ERRCAdmin016>:SubStatus<ES0001>:Host is not running in cluster.    
at Microsoft.ApplicationServer.Caching.AdminApi.CacheAdmin.StopHostHelper(IHostConfiguration hostConfig, Boolean clusterCommand, Int32 hostTimeout, Boolean graceful)    
at Microsoft.ApplicationServer.Caching.AdminApi.CacheAdmin.StopHost(String hostName, Int32 port, Nullable`1 hostTimeout, Boolean graceful)    
at Microsoft.ApplicationServer.Caching.Commands.StopAFCacheHostCommand.BeginProcessing()     -
-- End of inner exception stack trace ---    
at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)    
at Microsoft.SharePoint.DistributedCaching.Utilities.SPVelocityPowerShellWrapper.StopCacheHost(String provider, String connectionString, String hostName, Int32 cachePort, Boolean bGraceFul)'.
```

DSC logs:
```
VERBOSE: [2018-12-18 08:36:41Z] [VERBOSE] [SP]: LCM:  [ Start  Set      ]  
[[SPDistributedCacheService]EnableDistributedCache]
VERBOSE: [2018-12-18 08:36:41Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Setting the cache host 
information
VERBOSE: [2018-12-18 08:36:41Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Getting the cache host 
information
VERBOSE: [2018-12-18 08:36:41Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Executing using a provided 
credential and local PSSession as user contoso\spsetup
VERBOSE: [2018-12-18 08:36:45Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Adding the distributed 
cache to the server
VERBOSE: [2018-12-18 08:36:45Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Create a firewall rule for 
AppFabric
VERBOSE: [2018-12-18 08:36:45Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Executing using a provided 
credential and local PSSession as user contoso\spsetup
VERBOSE: [2018-12-18 08:36:51Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Firewall rule added
VERBOSE: [2018-12-18 08:36:51Z] Settings handler status to 'transitioning' 
(C:\Packages\Plugins\Microsoft.Powershell.DSC\2.77.0.0\Status\0.status)
VERBOSE: [2018-12-18 08:36:51Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Current state is 'Absent' 
and desired state is 'Present'
VERBOSE: [2018-12-18 08:36:51Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Enabling distributed cache 
service
VERBOSE: [2018-12-18 08:36:51Z] [VERBOSE] [SP]:                            
[[SPDistributedCacheService]EnableDistributedCache] Executing using a provided 
credential and local PSSession as user contoso\spsetup
VERBOSE: [2018-12-18 08:45:35Z] [ERROR] Exception calling "Deploy" with "0" 
argument(s): "Error occurred while performing the operation on host SP:22233 : 
ErrorCode<ERRCAdmin016>:SubStatus<ES0001>:Host is not running in cluster."
VERBOSE: [2018-12-18 08:45:35Z] Settings handler status to 'transitioning' 
(C:\Packages\Plugins\Microsoft.Powershell.DSC\2.77.0.0\Status\0.status)
VERBOSE: [2018-12-18 08:45:35Z] [VERBOSE] [SP]: LCM:  [ End    Set      ]  
[[SPDistributedCacheService]EnableDistributedCache]  in 534.0100 seconds.
VERBOSE: [2018-12-18 08:45:35Z] [ERROR] The PowerShell DSC resource 
'[SPDistributedCacheService]EnableDistributedCache' with SourceInfo 
'C:\Packages\Plugins\Microsoft.Powershell.DSC\2.77.0.0\DSCWork\ConfigureSPVM.0\
ConfigureSPVM.ps1::402::9::SPDistributedCacheService' threw one or more 
non-terminating errors while running the Set-TargetResource functionality. 
These errors are logged to the ETW channel called 
Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
```

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)

```powershell
SPFarm CreateSPFarm
{
	DatabaseServer            = $SQLAlias
	FarmConfigDatabaseName    = $SPDBPrefix + "Config"
	Passphrase                = $SPPassphraseCreds
	FarmAccount               = $SPFarmCredsQualified
	PsDscRunAsCredential      = $SPSetupCredsQualified
	AdminContentDatabaseName  = $SPDBPrefix + "AdminContent"
	CentralAdministrationPort = 5000
	RunCentralAdmin           = $true
	IsSingleInstance          = "Yes"
	Ensure                    = "Present"
	DependsOn                 = "[xScript]WaitForSQL"
}

SPManagedAccount CreateSPSvcManagedAccount
{
	AccountName          = $SPSvcCredsQualified.UserName
	Account              = $SPSvcCredsQualified
	PsDscRunAsCredential = $SPSetupCredsQualified
	DependsOn            = "[SPFarm]CreateSPFarm"
}

SPDistributedCacheService EnableDistributedCache
{
	Name                 = "AppFabricCachingService"
	CacheSizeInMB        = 2000
	CreateFirewallRules  = $true
	ServiceAccount       = $SPSvcCredsQualified.UserName
	InstallAccount       = $SPSetupCredsQualified
	Ensure               = "Present"
	DependsOn            = "[SPFarm]CreateSPFarm"
}
```

#### The operating system the target node is running

OsName               : Microsoft Windows Server 2016 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsBuildLabEx    : 14393.2551.amd64fre.rs1_release.181004-1309
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### Version of SharePoint that is used (e.g. SharePoint 2016)

SharePoint 2019 16.0.10337.12109

#### Version and build of PowerShell the target node is running

PSVersion                      5.1.14393.2515
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.2515
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

#### Version of the DSC module that was used ('dev' if using current dev branch)

SharePointDSC 3.1.0.0

**Comment by Yvand** [comment]
I investigated it and it fail when resource runs Stop-SPServiceInstance:

https://github.com/PowerShell/SharePointDsc/blob/3844111e1e98c1aa4483f4251845425264af8434/Modules/SharePointDsc/DSCResources/MSFT_SPDistributedCacheService/MSFT_SPDistributedCacheService.psm1#L249-L251

So I surrounded Stop-SPServiceInstance with try/catch like below and it fixed the issue:

```powershell
try
{
	Get-SPServiceInstance | Where-Object -FilterScript {
		$_.GetType().Name -eq "SPDistributedCacheServiceInstance"
	} | Stop-SPServiceInstance -Confirm:$false
}
catch
{
	# In SharePoint 2019, Stop-SPServiceInstance throws an exception if service is not running on the server
}
```

Is it a good fix?
If so, I will submit a PR soon with this change

**Comment by Yvand** [comment]
Solution I proposed above is not working, I continue to get the error after this, I need to continue the investigations...

**Comment by andikrueger** [comment]
A while back I created an issue on the same topic and it seems to be related: #786

**Comment by ykuijs** [comment]
Closed via #997
